# Project 7 -- Niharika Rai

**TA Help:** (for instance) John Smith, Alice Jones, etc., list names of any TAs who helped you

- For example: Help with figuring out how to write a function (describe the tasks that they helped you with)

**Collaboration:** My Friend in CS, My Uncle, Another Student, etc., list names of any other people who helped you

(describe the tasks that they helped you with)
- For example: helped figuring out how to load the dataset.
- Another example: helped debug error with my plot.

**Internet Resources:** Stack Exchange, Stack Overflow, etc.

(describe any information that you learned from internet resources, including the URLs)
- data frames in Pandas versus R from StackOverflow  https://stackoverflow.com/questions/8991709/why-were-pandas-merges-in-python-faster-than-data-table-merges-in-r-in-2012

**ChatGPT, Gemini, Claude, etc:** Any language models or generative AI chatbots that helped you.

(if you used any such tools, please tell us here)
- For example:  I asked ChatGPT how to define a new data frames
- Another example:  Gemini told me how to make a function for sorting my data

- ***Link to AI Chat History***: Here is the link: https://chatgpt.com/share/68ef3dca-c128-800d-8c6e-992237f27de5 Please share a link to your chat if you used AI (ex. ChatGPT Shared Links)
**OVERALL MESSAGE:** Any time that you used anything except your brain to solve the questions in these projects, you need to disclose such resources at the start of the project, with details about your usage of the tools.

**YOUR OWN WORK:** Even when you utilize other resources, do NOT just copy and paste.  Write all explanations in your own words, using several sentences in English, which are understandable and which you wrote (and did not just copy and paste).

## Question 1

In [1]:
%sql sqlite:////anvil/projects/tdm/data/lahman/lahman.db

In [2]:
%%sql
SELECT name FROM sqlite_master WHERE type='table';

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


name
teamsfranchises
teams
teamshalf
people
allstarfull
appearances
awardsmanagers
awardsplayers
awardssharemanagers
awardsshareplayers


In [3]:
%%sql
SELECT * FROM People
LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


playerID,birthYear,birthMonth,birthDay,birthCity,birthCountry,birthState,deathYear,deathMonth,deathDay,deathCountry,deathState,deathCity,nameFirst,nameLast,nameGiven,weight,height,bats,throws,debut,bbrefID,finalGame,retroID
aardsda01,1981,12,27,Denver,USA,CO,,,,,,,David,Aardsma,David Allan,215,75,R,R,2004-04-06,aardsda01,2015-08-23,aardd001
aaronha01,1934,2,5,Mobile,USA,AL,2021,1,22,USA,GA,Atlanta,Hank,Aaron,Henry Louis,180,72,R,R,1954-04-13,aaronha01,1976-10-03,aaroh101
aaronto01,1939,8,5,Mobile,USA,AL,1984,8,16,USA,GA,Atlanta,Tommie,Aaron,Tommie Lee,190,75,R,R,1962-04-10,aaronto01,1971-09-26,aarot101
aasedo01,1954,9,8,Orange,USA,CA,,,,,,,Don,Aase,Donald William,190,75,R,R,1977-07-26,aasedo01,1990-10-03,aased001
abadan01,1972,8,25,Palm Beach,USA,FL,,,,,,,Andy,Abad,Fausto Andres,184,73,L,L,2001-09-10,abadan01,2006-04-13,abada001


In [4]:
%%sql
SELECT * FROM Salaries
LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


yearID,teamID,lgID,playerID,salary
1985,ATL,NL,barkele01,870000.0
1985,ATL,NL,bedrost01,550000.0
1985,ATL,NL,benedbr01,545000.0
1985,ATL,NL,campri01,633333.0
1985,ATL,NL,ceronri01,625000.0


In [5]:
%%sql
SELECT * FROM Teams
LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


yearID,lgID,teamID,franchID,divID,teamRank,G,Ghome,W,L,DivWin,WCWin,LgWin,WSWin,R,AB,H,2B,3B,HR,BB,SO,SB,CS,HBP,SF,RA,ER,ERA,CG,SHO,SV,IPouts,HA,HRA,BBA,SOA,E,DP,FP,name,park,attendance,BPF,PPF,teamIDBR,teamIDlahman45,teamIDretro
1871,NA,BS1,BNA,,3,31,,20,10,,,N,,401,1372,426,70,37,3,60,19,73,16,,,303,109,3.55,22,1,3,828,367,2,42,23,243,24,0.834,Boston Red Stockings,South End Grounds I,,103,98,BOS,BS1,BS1
1871,NA,CH1,CNA,,2,28,,19,9,,,N,,302,1196,323,52,21,10,60,22,69,21,,,241,77,2.76,25,0,1,753,308,6,28,22,229,16,0.829,Chicago White Stockings,Union Base-Ball Grounds,,104,102,CHI,CH1,CH1
1871,NA,CL1,CFC,,8,29,,10,19,,,N,,249,1186,328,35,40,7,26,25,18,8,,,341,116,4.11,23,0,0,762,346,13,53,34,234,15,0.818,Cleveland Forest Citys,National Association Grounds,,96,100,CLE,CL1,CL1
1871,NA,FW1,KEK,,7,19,,7,12,,,N,,137,746,178,19,8,2,33,9,16,4,,,243,97,5.17,19,1,0,507,261,5,21,17,163,8,0.803,Fort Wayne Kekiongas,Hamilton Field,,101,107,KEK,FW1,FW1
1871,NA,NY2,NNA,,5,33,,16,17,,,N,,302,1404,403,43,21,1,33,15,46,15,,,313,121,3.72,32,1,0,879,373,7,42,22,235,14,0.84,New York Mutuals,Union Grounds (Brooklyn),,90,88,NYU,NY2,NY2


In [6]:
%%sql
SELECT
    p.nameFirst || ' ' || p.nameLast AS fullName,
    t.name AS teamName,
    s.yearID,
    s.salary
FROM Salaries s
JOIN People p ON p.playerID = s.playerID
JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
LIMIT 10;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


fullName,teamName,yearID,salary
Len Barker,Atlanta Braves,1985,870000.0
Steve Bedrosian,Atlanta Braves,1985,550000.0
Bruce Benedict,Atlanta Braves,1985,545000.0
Rick Camp,Atlanta Braves,1985,633333.0
Rick Cerone,Atlanta Braves,1985,625000.0
Chris Chambliss,Atlanta Braves,1985,800000.0
Jeff Dedmon,Atlanta Braves,1985,150000.0
Terry Forster,Atlanta Braves,1985,483333.0
Gene Garber,Atlanta Braves,1985,772000.0
Terry Harper,Atlanta Braves,1985,250000.0


In [7]:
%%sql
SELECT
    p.nameFirst || ' ' || p.nameLast AS fullName,
    t.name AS teamName,
    s.yearID,
    s.salary
FROM Salaries s
JOIN People p ON p.playerID = s.playerID
JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
ORDER BY s.salary DESC
LIMIT 10;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


fullName,teamName,yearID,salary
Alex Rodriguez,New York Yankees,2009,33000000.0
Alex Rodriguez,New York Yankees,2010,33000000.0
Clayton Kershaw,Los Angeles Dodgers,2016,33000000.0
Clayton Kershaw,Los Angeles Dodgers,2015,32571000.0
Alex Rodriguez,New York Yankees,2011,32000000.0
Zack Greinke,Arizona Diamondbacks,2016,31799030.0
Alex Rodriguez,New York Yankees,2012,30000000.0
David Price,Boston Red Sox,2016,30000000.0
Alex Rodriguez,New York Yankees,2013,29000000.0
Alex Rodriguez,New York Yankees,2008,28000000.0


In [8]:
%%sql
WITH joined_data AS (
    SELECT
        p.nameFirst || ' ' || p.nameLast AS fullName,
        t.name AS teamName,
        s.yearID,
        s.salary
    FROM Salaries s
    JOIN People p ON p.playerID = s.playerID
    JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
)
SELECT teamName, yearID, AVG(salary) AS avg_salary FROM joined_data GROUP BY teamName, yearID ORDER BY teamName, yearID;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


teamName,yearID,avg_salary
Anaheim Angels,1997,1004370.0645161291
Anaheim Angels,1998,1214147.0588235294
Anaheim Angels,1999,1384704.15
Anaheim Angels,2000,1715472.2333333334
Anaheim Angels,2001,1584505.5666666667
Anaheim Angels,2002,2204345.25
Anaheim Angels,2003,2927098.777777778
Anaheim Angels,2004,3723506.185185185
Arizona Diamondbacks,1998,898527.7777777778
Arizona Diamondbacks,1999,2020705.8529411764


1.1 I used the query given in the example but removed the part where we had to order in descending order (shown in 1.2 instead) and added a limit value of 10 because otherwise, my query was endlessly very long. 
1.2 I used the SQL query given in the example.
1.3 I used the SQL query given in the example and some part of queries from project 6 for reference to construct my code. I used ChatGPT to debug my code as I was getting syntax errors and the output for average salary of players on the New York Yankees in 2005 would not show up.  
1.4 Average Salary for the New York Yankees in 2005 was $8,011,800.65 (exact output: 8011800.653846154).  

## Question 2

In [9]:
%%sql
SELECT
    p.nameFirst || ' ' || p.nameLast AS fullName,
    t.name AS teamName,
    s.yearID,
    s.salary
FROM Salaries s
INNER JOIN People p ON p.playerID = s.playerID
INNER JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
LIMIT 10;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


fullName,teamName,yearID,salary
Len Barker,Atlanta Braves,1985,870000.0
Steve Bedrosian,Atlanta Braves,1985,550000.0
Bruce Benedict,Atlanta Braves,1985,545000.0
Rick Camp,Atlanta Braves,1985,633333.0
Rick Cerone,Atlanta Braves,1985,625000.0
Chris Chambliss,Atlanta Braves,1985,800000.0
Jeff Dedmon,Atlanta Braves,1985,150000.0
Terry Forster,Atlanta Braves,1985,483333.0
Gene Garber,Atlanta Braves,1985,772000.0
Terry Harper,Atlanta Braves,1985,250000.0


In [10]:
%%sql
SELECT
    p.nameFirst || ' ' || p.nameLast AS fullName,
    t.name AS teamName,
    s.yearID,
    s.salary
FROM Salaries s
LEFT JOIN People p ON p.playerID = s.playerID
LEFT JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
LIMIT 10;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


fullName,teamName,yearID,salary
Len Barker,Atlanta Braves,1985,870000.0
Steve Bedrosian,Atlanta Braves,1985,550000.0
Bruce Benedict,Atlanta Braves,1985,545000.0
Rick Camp,Atlanta Braves,1985,633333.0
Rick Cerone,Atlanta Braves,1985,625000.0
Chris Chambliss,Atlanta Braves,1985,800000.0
Jeff Dedmon,Atlanta Braves,1985,150000.0
Terry Forster,Atlanta Braves,1985,483333.0
Gene Garber,Atlanta Braves,1985,772000.0
Terry Harper,Atlanta Braves,1985,250000.0


In [11]:
%%sql
WITH joined_data AS (
    SELECT
        p.nameFirst || ' ' || p.nameLast AS fullName,
        t.name AS teamName,
        s.yearID,
        s.salary
    FROM Salaries s
    INNER JOIN People p ON p.playerID = s.playerID
    INNER JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
)
SELECT COUNT(*) AS total_rows FROM joined_data;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


total_rows
26428


In [12]:
%%sql
WITH joined_data AS (
    SELECT
        p.nameFirst || ' ' || p.nameLast AS fullName,
        t.name AS teamName,
        s.yearID,
        s.salary
    FROM Salaries s
    LEFT JOIN People p ON p.playerID = s.playerID
    LEFT JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
)
SELECT COUNT(*) AS total_rows FROM joined_data;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


total_rows
26428


In [16]:
%%sql
WITH joined_data AS (
    SELECT
        p.nameFirst || ' ' || p.nameLast AS fullName,
        t.name AS teamName,
        s.yearID,
        s.salary
    FROM Salaries s
    LEFT JOIN People p ON p.playerID = s.playerID
    LEFT JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
)
SELECT COUNT(*) AS null_total_rows FROM joined_data WHERE fullName IS NULL OR teamName IS NULL OR salary IS NULL;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


null_total_rows
0


In [17]:
%%sql
WITH joined_data AS (
    SELECT
        p.nameFirst || ' ' || p.nameLast AS fullName,
        t.name AS teamName,
        s.yearID,
        s.salary
    FROM Salaries s
    INNER JOIN People p ON p.playerID = s.playerID
    INNER JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
)
SELECT COUNT(*) AS null_total_rows FROM joined_data WHERE fullName IS NULL OR teamName IS NULL OR salary IS NULL;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


null_total_rows
0


2.1 I used the query given in the example but removed the part where we had to order the salary in descending order as that is not specified by the question and added a limit value of 10 because otherwise, my query was endlessly long. 
2.2 I used the query given in the example but modified it the same way as I had done for 2.1. 
2.3 Both the inner and left join produced query results look the same, but only for this dataset as it is well-structured and is also stated in the question. The results of both queries have the same number of rows (26428) and they do not have any null values either. I used ChatGPT to get an idea of how to detect null values in SQL queries and found out about the 'IS NULL' condition that can be applied using the 'WHERE' condition. I counted the rows with null values to figure out if there any missing values in the results created using inner and left joins to support my reasoning.    

## Question 3

In [18]:
%%sql 
SELECT
    t.name AS teamName,
    s.yearID,
    AVG(s.salary) AS avgSalary
FROM Salaries s
JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
GROUP BY t.name, s.yearID;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


teamName,yearID,avgSalary
Anaheim Angels,1997,1004370.0645161291
Anaheim Angels,1998,1214147.0588235294
Anaheim Angels,1999,1384704.15
Anaheim Angels,2000,1715472.2333333334
Anaheim Angels,2001,1584505.5666666667
Anaheim Angels,2002,2204345.25
Anaheim Angels,2003,2927098.777777778
Anaheim Angels,2004,3723506.185185185
Arizona Diamondbacks,1998,898527.7777777778
Arizona Diamondbacks,1999,2020705.8529411764


In [20]:
%%sql
SELECT
    t.name AS teamName,
    s.yearID,
    AVG(s.salary) AS avgSalary
FROM Salaries s
JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
GROUP BY t.name, s.yearID
HAVING AVG(s.salary) > 7500000;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


teamName,yearID,avgSalary
New York Yankees,2005,8011800.653846154
New York Yankees,2009,7748045.730769231
New York Yankees,2010,8253335.56
New York Yankees,2014,8230996.125
New York Yankees,2016,7689579.0344827585


3.1 I used the sample query given in the question and modified it to remove the 'HAVING' and 'LIMIT' clause. My result is a single SQL query returning teamName, yearID, and avgSalary ordered by team and year.
3.2 I used the query that I wrote for 3.1 and added the 'HAVING' clause to filter for teams with an average salary greater than 7.5 million.
3.3 Based on the output in 3.2, only New York Yankees meet this criteria for years 2005, 2009, 2010, 2014 and 2016. Hence, only 1 team met this criteria for a total of 5 times. 

## Question 4

In [21]:
%%sql 
SELECT
    t.name AS teamName,
    s.yearID,
    AVG(s.salary) AS avgSalary,
    COUNT(s.playerID) AS numPlayers
FROM Salaries s
JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
GROUP BY t.name, s.yearID;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


teamName,yearID,avgSalary,numPlayers
Anaheim Angels,1997,1004370.0645161291,31
Anaheim Angels,1998,1214147.0588235294,34
Anaheim Angels,1999,1384704.15,40
Anaheim Angels,2000,1715472.2333333334,30
Anaheim Angels,2001,1584505.5666666667,30
Anaheim Angels,2002,2204345.25,28
Anaheim Angels,2003,2927098.777777778,27
Anaheim Angels,2004,3723506.185185185,27
Arizona Diamondbacks,1998,898527.7777777778,36
Arizona Diamondbacks,1999,2020705.8529411764,34


In [24]:
%%sql 
SELECT
    t.name AS teamName,
    s.yearID,
    AVG(s.salary) AS avgSalary,
    COUNT(s.playerID) AS numPlayers
FROM Salaries s
JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
GROUP BY t.name, s.yearID
HAVING AVG(s.salary) > 6800000 AND COUNT(s.playerID) > 20;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


teamName,yearID,avgSalary,numPlayers
Detroit Tigers,2015,6891390.0,25
Los Angeles Dodgers,2013,6980068.625,32
Los Angeles Dodgers,2015,7441103.448275862,29
New York Yankees,2005,8011800.653846154,26
New York Yankees,2006,6952252.821428572,28
New York Yankees,2008,6929892.966666667,30
New York Yankees,2009,7748045.730769231,26
New York Yankees,2010,8253335.56,25
New York Yankees,2011,6975000.9655172415,29
New York Yankees,2013,7483189.870967742,31


In [34]:
%%sql 
SELECT
    t.name AS teamName,
    s.yearID,
    AVG(s.salary) AS avgSalary,
    COUNT(s.playerID) AS numPlayers
FROM Salaries s
JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
WHERE s.yearID BETWEEN 2010 AND 2015
GROUP BY t.name, s.yearID
HAVING AVG(s.salary) > 6800000 AND COUNT(s.playerID) > 20;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


teamName,yearID,avgSalary,numPlayers
Detroit Tigers,2015,6891390.0,25
Los Angeles Dodgers,2013,6980068.625,32
Los Angeles Dodgers,2015,7441103.448275862,29
New York Yankees,2010,8253335.56,25
New York Yankees,2011,6975000.9655172415,29
New York Yankees,2013,7483189.870967742,31
New York Yankees,2014,8230996.125,24
New York Yankees,2015,7336274.379310345,29


4.1 I used the query given in the example and modified it to remove the 'HAVING', 'ORDER BY' and 'LIMIT' clauses. 
4.2 I used the query given in the example and modified it to add the 'HAVING' clause with given numerical conditions. 
4.3 I used the query given in the example and modified it to add the 'WHERE' clause and 'BETWEEN' operator to meet the year conditions. I ran into syntax error and used ChatGPT to debug my code. I realized that the 'WHERE' clause must always come before the 'GROUP BY' and 'HAVING' clauses for the query to work. 
4.4 Based on the output in 4.3, only 3 teams meet the given criteria. These teams are Detroit Tigers, Los Angeles Dodgers and the New York Yankees. 

## Question 5

In [56]:
%%sql
WITH payroll AS (
    SELECT
        teamID,
        yearID,
        SUM(salary) AS totalPayroll
    FROM Salaries
    GROUP BY teamID, yearID
),
homeruns AS (
    SELECT
        teamID,
        yearID,
        MAX(HR) AS maxHR
    FROM Teams
    GROUP BY teamID, yearID
)
SELECT
    p.teamID,
    p.yearID,
    p.totalPayroll,
    h.maxHR
FROM payroll p
JOIN homeruns h
    ON p.teamID = h.teamID AND p.yearID = h.yearID
WHERE p.totalPayroll > 150000000 AND h.maxHR >= 30
ORDER BY p.totalPayroll ASC
LIMIT 5;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


teamID,yearID,totalPayroll,maxHR
BOS,2013,151530000.0,178
NYA,2003,152749814.0,230
DET,2014,152855500.0,155
CHN,2016,154067668.0,199
LAA,2012,154485166.0,187


5.1 I used the sample query in the question and added the HR aggregration, 'JOIN', 'WHERE', 'ORDER BY' and 'LIMIT' clauses. I was confused on how to join two queries where both have been constructed using the 'WITH' clause and needed ChatGPT's assisstance. I learnt that both queries could be defined under the same 'WITH' clause but separated by a comma. Then, the 'JOIN' clause can be applied once they are aggregated. 
5.2 Based on the output for the query created in 5.1, the top 5 cheapest team-seasons that meet the given criteria are BOS, NYA, DET, CHN, and LAA. 

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.

https://www.purdue.edu/odos/osrr/honor-pledge/
